src/java.desktop/share/classes/sun/print/PrintJobDelegate.java

```diff
@@ -0,0 +1,1233 @@
+
+        Graphics printGraphics = null;
+
+        synchronized (this) {
+            ++pageIndex;
+
+            // Thread should not be created after end has been called.
+            // One way to detect this is if any of the graphics queue
+            //  has been closed.
+            if (pageIndex == 0 && !graphicsToBeDrawn.isClosed()) {
+
+            /* We start a thread on which the PrinterJob will run.
+             * The PrinterJob will ask for pages on that thread
+             * and will use a message queue to fulfill the application's
+             * requests for a Graphics on the application's
+             * thread.
+             */
+
+                startPrinterJobThread();
+
+            }
+            notify();
+        }
+
+        /* If the application has already been handed back
+         * a graphics then we need to put that graphics into
+         * the drawn queue so that the PrinterJob thread can
+         * return to the print system.
+         */
+        if (currentGraphics != null) {
+            graphicsDrawn.append(currentGraphics);
+            currentGraphics = null;
+        }
+
+        /* We'll block here until a new graphics becomes
+         * available.
+         */
+
+        currentGraphics = graphicsToBeDrawn.pop();
+
+        if (currentGraphics instanceof PeekGraphics) {
+            ( (PeekGraphics) currentGraphics).setAWTDrawingOnly();
+            graphicsDrawn.append(currentGraphics);
+            currentGraphics = graphicsToBeDrawn.pop();
+        }
+
+
+        if (currentGraphics != null) {
+
+            /* In the PrintJob API, the origin is at the upper-
+             * left of the imageable area when using the new "printable"
+             * origin attribute, otherwise its the physical origin (for
+             * backwards compatibility. We emulate this by creating
+             * a PageFormat which matches and then performing the
+             * translate to the origin. This is a no-op if physical
+             * origin is specified.
+             */
+            currentGraphics.translate(pageFormat.getImageableX(),
+                                      pageFormat.getImageableY());
+
+            /* Scale to accommodate AWT's notion of printer resolution */
+            double awtScale = 72.0/getPageResolutionInternal();
+            currentGraphics.scale(awtScale, awtScale);
+
+            /* The caller wants a Graphics instance but we do
+             * not want them to make 2D calls. We can't hand
+             * back a Graphics2D. The returned Graphics also
+             * needs to implement PrintGraphics, so we wrap
+             * the Graphics2D instance. The PrintJob API has
+             * the application dispose of the Graphics so
+             * we create a copy of the one returned by PrinterJob.
+             */
+            //printGraphics = new ProxyPrintGraphics(currentGraphics.create(), this);
+            printGraphics =  currentGraphics.create();
```

Comment:
Do we need the comment on line 823?

> The returned Graphics also  to implement PrintGraphics, so we  wrap the Graphics2D instance.

That doesn't seem true anymore.

Reply:
Moving part of this comment back over to PrintJob2D.